pkg/controller/master-controller-manager/master-constraint-controller/controller.go

```diff
@@ -0,0 +1,213 @@
+	}
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) syncAllSeeds(ctx context.Context, log *zap.SugaredLogger, constraint *kubermaticv1.Constraint, action func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint) error) error {
+	for seedName, seedClient := range r.seedClients {
+
+		log := log.With("seed", seedName)
+
+		log.Debug("Reconciling constraint with seed")
+
+		err := action(seedClient, constraint)
+		if err != nil {
+			return fmt.Errorf("failed syncing constraint %s for seed %s: %w", constraint.Name, seedName, err)
+		}
+		log.Debug("Reconciled constraint with seed")
+	}
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+
+	constraint := &kubermaticv1.Constraint{}
+	if err := r.masterClient.Get(ctx, request.NamespacedName, constraint); err != nil {
+		if controllerutil.IsCacheNotStarted(err) {
+			return err
+		}
+
+		if kerrors.IsNotFound(err) {
+			log.Debug("constraint not found, returning")
+			return nil
+		}
+
+		return fmt.Errorf("failed to get constraint %s: %v", request.Name, err)
+	}
+
+	// handling deletion
+	if !constraint.DeletionTimestamp.IsZero() {
+
+		if err := r.handleDeletion(ctx, log, constraint); err != nil {
+			return fmt.Errorf("handling deletion of constraint: %v", err)
+		}
+		return nil
+	}
+
+	if !kuberneteshelper.HasFinalizer(constraint, kubermaticapiv1.GatekeeperSeedConstraintCleanupFinalizer) {
+		oldconstraint := constraint.DeepCopy()
+		kuberneteshelper.AddFinalizer(constraint, kubermaticapiv1.GatekeeperSeedConstraintCleanupFinalizer)
+		if err := r.masterClient.Patch(ctx, constraint, ctrlruntimeclient.MergeFrom(oldconstraint)); err != nil {
+			return fmt.Errorf("failed to set constraint finalizer %s: %v", constraint.Name, err)
+		}
+	}
+
+	constraintCreatorGetters := []reconciling.NamedKubermaticV1ConstraintCreatorGetter{
+		constraintCreatorGetter(constraint),
+	}
+
+	return r.syncAllSeeds(ctx, log, constraint, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint) error {
+		return reconciling.ReconcileKubermaticV1Constraints(ctx, constraintCreatorGetters, r.namespace, seedClient)
+	})
+}
+
+func (r *reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, constraint *kubermaticv1.Constraint) error {
+
+	// if finalizer not set to master Constraint then return
+	if !kuberneteshelper.HasFinalizer(constraint, kubermaticapiv1.GatekeeperSeedConstraintCleanupFinalizer) {
+		return nil
+	}
+
+	err := r.syncAllSeeds(ctx, log, constraint, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint) error {
+		err := seedClient.Delete(ctx, &kubermaticv1.Constraint{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      constraint.Name,
+				Namespace: constraint.Namespace,
+			},
+		})
+
+		if kerrors.IsNotFound(err) {
```

Comment:
you can use `IgnoreNotFound(err)` from `"sigs.k8s.io/controller-runtime/pkg/client"`: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/interfaces.go#L140